Move students: Add redux, UI tweaks, refactoring

maddiedierker opened the pull request:
This is the third in a series of PRs to rewrite 'move students' in React. Previous PR [here](https://github.com/code-dot-org/code-dot-org/pull/22099), spec [here](https://docs.google.com/document/d/18qEq8lQIvPrpMaWwCeAPsCXAPvYwVS7NYoMIt7_swvY/edit?usp=sharing), and Axisoft ticket [here](https://codeorg.axosoft.com/viewitem?id=1439&type=features&force_use_number=true).

**These changes include:**
- Add `transferStudents` to redux
- UI tweaks: disable 'Move students' button until the following conditions are met (images at very bottom of comment to illustrate):
  1. at least 1 student is chosen
  2. a section is chosen from the dropdown (if moving to a section owned by the same teacher), OR a section code is input (if moving to a section owned by a different teacher)
- Refactor state to `manageStudentsRedux` that previously lived in the `MoveStudents` component (includes some renaming for clarity in context of reducer rather than component)
- Refactor `MoveStudents` unit tests to test event simulation rather than output of helper methods (these are basically totally rewritten, so I'm sorry that the diff is confusing)

**Conditions to enable 'Move students' button:**
<img width="832" alt="screen shot 2018-05-01 at 3 11 04 pm" src="https://user-images.githubusercontent.com/9812299/39496114-02d9b1ac-4d52-11e8-9eb3-268e1b6718df.png">
<img width="831" alt="screen shot 2018-05-01 at 3 10 48 pm" src="https://user-images.githubusercontent.com/9812299/39496116-03e744ce-4d52-11e8-94be-12a2d8f38d0b.png">

**What's left?**
- Add success/error/information notifications after moving students
- Sort students alphabetically on initial modal open
- Style tweaks
- Update storybook
- Incorporate `/sections/transfers` endpoint changes and UI changes once new 'move students' is ready
- Remove old code
- & feedback!

caleybrock commented on apps/src/templates/manageStudents/MoveStudents.jsx:
```diff
@@ -10,14 +10,13 @@ import Button from '../Button';
 import BaseDialog from '../BaseDialog';
 import DialogFooter from "../teacherDashboard/DialogFooter";
 import {sectionsNameAndId} from '@cdo/apps/templates/teacherDashboard/teacherSectionsRedux';
+import {updateStudentTransfer, transferStudents, OTHER_TEACHER, COPY_STUDENTS, blankStudentTransfer} from './manageStudentsRedux';
```
nit: This line is too long, break up into multiple.

caleybrock commented on apps/src/templates/manageStudents/MoveStudents.jsx:
```diff
@@ -174,13 +179,15 @@ class MoveStudents extends Component {
         header: {
           label: i18n.name(),
           props: {
+            id: 'name-header',
```
nit: when these are for testing only, we typically prepend with `uitest` like this: https://github.com/code-dot-org/code-dot-org/blob/9ea4132bda304653fb4dccf29a90a993ca3d3ce9/apps/src/templates/studioHomepages/SetUpSections.jsx#L22

caleybrock commented on apps/src/templates/manageStudents/manageStudentsRedux.js:
```diff
@@ -209,6 +235,34 @@ export const addMultipleAddRows = (studentNames) => {
+export const transferStudents = () => {
+  return (dispatch, getState) => {
+    const state = getState();
+    const currentSectionCode = sectionCode(state, state.manageStudents.sectionId);
```
Given our original concern, might be good to have a comment here pointing out we are pulling information from another part of the redux state tree.

davidsbailey commented on apps/src/templates/manageStudents/manageStudentsRedux.js:
```diff
@@ -16,6 +17,27 @@ export const RowType = {
+// OTHER_TEACHER - value determines whether students will be moved to a section owned by a different teacher
+// COPY_STUDENTS - value determines whether students will be copied to the new section or moved (and subsequently removed from current section)
+export const OTHER_TEACHER = "otherTeacher";
+export const COPY_STUDENTS = "copy";
```
nit: life might be easier in some future debugging session if this value is "copyStudents" rather than just "copy"

davidsbailey replied:
Also, it would be more consistent with our style to write the comment closer to each constant, e.g.:
```
// Constants around moving students to another section.

// whether students will be moved to a section owned by a different teacher
export const OTHER_TEACHER = "otherTeacher";

// ...
export const COPY_STUDENTS = "copyStudents";
```

maddiedierker commented:
all good points, @caleybrock! i've updated my PR with your feedback, and fixed the storybook for this component since it caused some tests in circleci to fail

davidsbailey commented on apps/src/templates/manageStudents/manageStudentsRedux.js:
```diff
@@ -16,6 +17,27 @@ export const RowType = {
+export const COPY_STUDENTS = "copy";
+
+/** Initial state for manageStudents.transferData redux store.
+  * studentIds - student ids selected to be moved to another section
```
For multi-line comments we follow the style shown here, with comment starting on the next line, and aligning with the first `*`, and ending with ` */` like this:
```
/**
 * Initial state...
 * ...
 */
```
I know it's no fun to receive arbitrary style corrections, but we believe it'll make our code more readable if we follow our style guide closely.

davidsbailey commented on apps/test/unit/templates/manageStudents/MoveStudentsTest.js:
```diff
@@ -1,98 +1,137 @@
 import {expect} from '../../../util/configuredChai';
-import {UnconnectedMoveStudents as MoveStudents, DEFAULT_STATE} from '@cdo/apps/templates/manageStudents/MoveStudents';
+import sinon from 'sinon';
+import {blankStudentTransfer} from '@cdo/apps/templates/manageStudents/manageStudentsRedux';
```
nice reuse!

maddiedierker commented:
@davidsbailey thank you! i agree about having a cohesive style, so i don't mind stylistic nits at all 😃

caleybrock commented on apps/src/templates/manageStudents/MoveStudents.jsx:
```diff
@@ -357,5 +396,13 @@ class MoveStudents extends Component {
+  updateStudentTransfer(transferData) {
+    dispatch(updateStudentTransfer(transferData));
+  },
+  transferStudents(studentIds, currentSectionCode, newSectionCode, copyStudents) {
```
I think these params aren't needed.
